File: HW2/PathSums/solutions/bfs_soln.cpp
#include <iostream>
#include <tuple>
#include <vector>

using namespace std;
 
int n, k;
vector<tuple<int, int, int>> tree;

int bfs(int head, int sum) {
    if (head == -1) return 0;
    int value = get<0>(tree[head]);
    return (value == sum) + bfs(get<1>(tree[head]), sum - value) + bfs(get<2>(tree[head]), sum - value);
}

int main() {
    cin >> n;
    
    tree = vector<tuple<int, int, int>>(n);
    for (int i = 0; i < n; i++) {
        cin >> get<0>(tree[i]) >> get<1>(tree[i]) >> get<2>(tree[i]);
    }
    
    cin >> k;
    
    int num_sums = 0;
    for (int i = 0; i < n; i++) {
        num_sums += bfs(i, k);
    }
    cout << num_sums << endl;
    
    return 0;
}

File: HW2/PathSums/solutions/bfs_soln.py
n = int(input())
tree = [tuple(map(int, input().split())) for _ in range(n)]
k = int(input())

def bfs(head, sum):
    if head == -1:
        return 0
    value = tree[head][0]
    return (value == sum) + bfs(tree[head][1], sum - value) + bfs(tree[head][2], sum - value)

num_sums = 0
for i in range(n):
    num_sums += bfs(i, k)

print(num_sums)

